pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java

```diff
@@ -568,7 +574,8 @@ public void setRuleSetFactoryCompatibilityEnabled(boolean ruleSetFactoryCompatib
      */
     public AnalysisCache getAnalysisCache() {
         // Make sure we are not null
-        if (analysisCache == null) {
+        if (analysisCache == null || isIgnoreIncrementalAnalysis() && isAnalysisCacheFunctional()) {
+            // sets a noop cache
             setAnalysisCache(null);
```

Comment:
passing null here was a lousy hack to avoid double logging of the warning (as it was in the constructor). I'd take this  chance to initialize the `analysisCache` right upon definition and skip this logic in the getter altogether.

Reply:
I think the best I can do is replace this line with `setAnalysisCache(new NoopAnalysisCache())`... The difficulty is that I'd like the result of `getAnalysisCache` to be independent of the order of calls to `setIgnoreIncrementalAnalysis` and `setAnalysisCache`. 

I can't come up with anything cleaner than guarding the result in the getter... apart from making PMDConfiguration immutable and using a builder pattern, which would be nice, but would significantly break its interface

Reply:
`setAnalysisCache` could be private too, what do you think about deprecating its public usage?

Reply:
@oowekyala I don't know... moving all logic into this class seems wrong if we decide to continue building cache storage alternatives (ie: database, a network file, etc.)... I've no concrete plans for this now, but seeing the direction Gradle's build cache is heading is interesting as an idea source.